File: week1/python/2_tree_height.py
import sys
sys.setrecursionlimit(10**7)


class Tree:
    def __init__(self, n, parents):
        self.n = n
        self.parents = parents

    def height(self):
        heights = [0 for _ in range(self.n)]
        for vertex in range(self.n):
            if(heights[vertex] != 0):
                continue
            height = 0
            i = vertex
            while i != -1:
                if(heights[i] != 0):
                    height += heights[i]
                    break
                height += 1
                i = self.parents[i]
            i = vertex
            while i != -1:
                if heights[i] != 0:
                    break
                heights[i] = height
                height -= 1
                i = self.parents[i]
        return max(heights)


n = int(input())
parents = list(map(int, input().split()))

tree = Tree(n, parents)

print(tree.height())

File: week1/c++/2_tree_height.cpp
#include <iostream>
#include <algorithm>
#include <vector>

using namespace std;

int main()
{
    int n;
    vector<int> parents;
    cin >> n;
    for (int i = 0; i < n; ++i)
    {
        int inp;
        cin >> inp;
        parents.push_back(inp);
    };

    vector<int> heights;
    for (int i = 0; i < n; ++i)
    {
        heights.push_back(0);
    };

    int maxheight = 0;
    for (int vertex = 0; vertex < n; ++vertex)
    {
        if (heights[vertex] != 0)
        {
            continue;
        }
        int i = vertex;
        int height = 0;

        while (i != -1)
        {
            if (heights[i] != 0)
            {
                height += heights[i];
                break;
            }
            height++;
            i = parents[i];
        };

        maxheight = max(height, maxheight);
        i = vertex;

        while (i != -1)
        {
            if (heights[i] != 0)
            {
                break;
            }
            heights[i] = height;
            height--;
            i = parents[i];
        };
    };
    cout << maxheight << endl;
    return 0;
}